api/document/serializers/content.py

```diff
@@ -102,59 +102,91 @@ def to_internal_value(self, data: PrimitiveDict) -> PrimitiveDict:
         return serializer.to_internal_value(data)
 
 
-class BaseAnnotationSerializer(serializers.Serializer):
-    # This determines whether the annotation is a leaf of our
-    # content tree.
-    IS_LEAF = False
+class NestableAnnotationField(serializers.Field):
+    def get_attribute(self,
+                      instance: NestableAnnotation) -> NestableAnnotation:
+        return instance
 
-    content_type = serializers.SerializerMethodField()
-    inlines = serializers.SerializerMethodField()
-    text = serializers.SerializerMethodField()
 
-    @property
-    def CONTENT_TYPE(self):     # noqa; this is a constant
-        raise NotImplementedError('Unknown annotation type')
+class InlinesField(NestableAnnotationField):
+    def __init__(self, is_leaf_node: bool):
+        super().__init__()
+        self.is_leaf_node = is_leaf_node
+
+    def to_representation(self,
+                          instance: NestableAnnotation) -> List[PrimitiveDict]:
+        if self.is_leaf_node:
+            return []
+        return NestedAnnotationSerializer(
+            instance.children, context=self.context, many=True).data
+
+    def to_internal_value(self,
+                          data: List[PrimitiveDict]) -> List[PrimitiveDict]:
+        if not self.is_leaf_node:
+            serializer = NestedAnnotationSerializer()
+            return [
+                serializer.to_internal_value(item) for item in data
+            ]
+        elif data:
+            raise ValidationError('leaf nodes cannot contain nested content')
+        return []
+
+
+class TextField(NestableAnnotationField):
+    def __init__(self, is_leaf_node: bool):
+        super().__init__()
+        self.is_leaf_node = is_leaf_node
 
     @property
     def doc_node_text(self):
         return self.context['cursor'].text
 
+    def to_representation(self, instance: NestableAnnotation) -> str:
+        return self.doc_node_text[instance.start:instance.end]
+
+    def to_internal_value(self, data: str) -> str:
+        if self.is_leaf_node:
+            serializer = serializers.CharField(trim_whitespace=False)
+            return serializer.to_internal_value(data)
+
+        # If this isn't a leaf node, we're not going to complain,
+        # because we do fill out this value on non-leaf nodes during
+        # serialization, and we want clients to be able to modify
+        # serialized responses and send them back without having to
+        # do too much work.
+        #
+        # However, that said, we *will* throw away the text value on
+        # non-leaf nodes during deserialization, to make sure it's not
+        # accidentally used by anything further down the pipeline.
+        return ''
```

Comment:
Another option would be to remove the text field on non-leaf nodes. That'd complicate our logic (in the UI and during serialization) a bit more, but if it makes a better API, it might be worth it. If we do something like that, it probably'll make sense to split the serializer implementations into different classes. Something to ponder.